### PR TITLE
IMAP and POP3 specific code moved to backend classes.

### PR DESCRIPTION
**Pull request by tikank**
This is first part of https://github.com/pulb/mailnag/issues/21. I moved IMAP and POP3 code to backend classes. The code logic should be same as before. I hope I didn't break anything. I changed some method names so that backend interface would not be so protocol specific.

### CODE REVIEW

**Comment by pulb** [comment]
Many thanks! I'll review this ASAP.

File: Mailnag/common/accounts.py [hunk 1]
```diff
@@ -63,16 +63,20 @@ def __init__(self, enabled = False, name = '', user = '', \
 		self.imap = imap # bool		
 		self.idle = idle # bool
 		self.folders = folders
+		self.backend = backend
 		self._conn = None
```
Comment (tikank) [inline Mailnag/common/accounts.py]: Not needed here. Connection is handled inside backends.

File: Mailnag/backends/pop3.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		lst = []
+		return lst
+
+
+	def notify_next_change(self, callback=None, timeout=None):
+		raise NotImplemented("POP3 does not support notifications")
+
+
+	def cancel_notifications(self):
+		raise NotImplemented("POP3 does not support notifications")
```
Comment (tikank) [inline Mailnag/backends/pop3.py]: Should be NotImplementedError.

**Comment by pulb** [comment]
Looks very promising and clean so far, though I still didn't find the time to review it completely.
A few first notes:
- "Backend" sounds a little too abstract to me, what about "MailBackend"?
- I know python supports duck typing, but I don't like it because you can't see a types interface at a glance. Can you add an interface (e.g. IMailBackend) to Mailnag/(Mail)Backends and inherit Pop3MailBackend and ImapMailBackend from it (analogous to Mailnag/plugins)?
- I think Account.open/close should be renamed into connect/disconnet
- Don't add other authors to files you added (e.g. backends/*.py) and add yourself to all files you modified 

All in all much cleaner and much more maintainable than before :+1:
I'll try to finish the review this weekend.

**Comment by tikank** [comment]
Good comments. Thanks. I will make changes.

I agree, plain backend is too vague. Maybe MailboxBackend (somehow I associate "mail" as a single mail). And interface definition would indeed be good.

About open/close vs. connect/disconnect. I was thinking ahead about local mailboxes. There is not actually any kind of connection with them, but they can be opened and closed. I can also think opening and closing IMAP or POP3 mailboxes. That is why i named them open and close. I leave them now this way, but I can rename them later if necessary.

Account class is now very thin and does almost nothing, because I moved most of the code to backends. I would like to do something about it later. Maybe rename it Mailbox (so that Mailbox uses MailboxBackends to access mailboxes). Or maybe remove it if it does nothing. Currently it is used so many places and I don't know yet if it has some role when local mailboxes are implemented. So I leave it now as is.

**Comment by tikank** [comment]
(A sidenote: whatever will be the fix for https://github.com/pulb/mailnag/issues/133, the fix should be made here also as all relevant code has been moved.)

**Comment by pulb** [comment]
Ok, I see. 
Correcktion regarding my authors comment: only add other authors to new files if you copied code from existing files.

**Comment by pulb** [comment]
Pheeeew.. just finished the review :grimacing: 
Awesome job! I'm going to push some fixes, so please don't forget to pull before applying further changes.

**Comment by pulb** [comment]
> Account class is now very thin and does almost nothing, because I moved most of the code to backends. I would like to do something about it later. Maybe rename it Mailbox (so that Mailbox uses MailboxBackends to access mailboxes). Or maybe remove it if it does nothing. Currently it is used so many places and I don't know yet if it has some role when local mailboxes are implemented. So I leave it now as is.

I think it can't be removed as you need a generic account data structure for all account (mailbox) types, independent of specifc implementations.

**Comment by tikank** [comment]
Super! Thanks! That was indeed big change. Next one will be smaller. :-)

**Comment by pulb** [comment]
Well, thank *you* :)

Though I'm still undecided whether Account.open/close are the right terms. 
IMHO opening/closing a local mailbox is some kind of corner case, in generall you're connecting/disconnecting remote mailboxes.

**Comment by pulb** [comment]
Oh and before changing something UI specific, I think I should merge the headerbars branch.

**Comment by tikank** [comment]
Only UI specific I am changing now is that "todo" part I added to accountdialog.py in commit https://github.com/pulb/mailnag/pull/132/commits/b6cbf7b9dec2b3cfd806bfaaeef4a021ae00d010, but I can fix that again if merge causes conflict.

**Comment by tikank** [comment]
I am getting feeling that open/close/connect/disconnect are implementation details and maybe they should not be in Account of MailboxBackend interfaces at all. Instead backend implementation could do those operations automatically when needed. IMAP backend could connect and disconnect if needed when list_messages or request_folders are called, and keep connection open when someone is waiting notifications**.

My work-in-progress mbox implementation does not need open and close methods because they are done inside list_messages (or request_folders if mbox would have folders), and there is no reason to keep mailbox open outside of those operations.

** notify_next_change matches now current IMAP implementation and I named it so when I moved code to backend. Maybe notify_next_change should actually be notify_changes (or register_change_callback or something like that) and it should keep notifying changes until notifications are canceled.

**Comment by pulb** [comment]
Yes, this actually a very good point - open/close should be removed completely and list_messages, request_folders and notify_changes should all use their own connection. notify_changes can even use multiple connections (one per monitored folder) and cancel_notification should close those (and possible call noop before). Mabye the code in idlers.py can be moved to notify_changes completely. I'll try to find time this week to implement this.

**Pull request merged**